server/implementation/src/main/java/io/smallrye/graphql/scalar/custom/IntCoercing.java

```diff
@@ -0,0 +1,83 @@
+package io.smallrye.graphql.scalar.custom;
+
+import java.math.BigInteger;
+
+import graphql.language.IntValue;
+import graphql.language.Value;
+import graphql.schema.Coercing;
+import graphql.schema.CoercingParseLiteralException;
+import graphql.schema.CoercingParseValueException;
+import graphql.schema.CoercingSerializeException;
+import io.smallrye.graphql.spi.ClassloadingService;
+
+public class IntCoercing implements Coercing<Object, BigInteger> {
+
+    private final Class<? extends CustomIntScalar> customScalarClass;
+
+    public IntCoercing(String customScalarClass) {
+        ClassloadingService classloadingService = ClassloadingService.get();
+        this.customScalarClass = (Class<? extends CustomIntScalar>) classloadingService.loadClass(customScalarClass);
+    }
+
+    private CustomIntScalar newInstance(BigInteger graphqlPrimitiveValue) {
+        try {
+            return graphqlPrimitiveValue == null ? null
+                    : customScalarClass.getConstructor(BigInteger.class).newInstance(graphqlPrimitiveValue);
```

Comment:
AFAIK this requires that the user always creates a public constructor with this specific signature. That's not the the nicest requirement, but yeah I don't have a better idea that's not unnecessarily complicated.
 If we really do it, I think it would be nice to scan during build-time to verify that this constructor exists and throw an error right away, instead of letting it go all the way here (until a request at runtime). 
That could happen inside `CustomScalarCreator.create()` before you actually create the `CustomScalarType`instance - here you have the `ClassInfo` so you can check that the class has the appropriate constructor.

Reply:
Good idea! I also moved the definition of the primitive type into this class as well.